datadriven/src/test_data_reader.rs

```diff
@@ -0,0 +1,156 @@
+            // build-scalar \
+            // vars(int)
+            while line.ends_with('\\') {
+                line.pop();
+
+                let l = self
+                    .scanner
+                    .next()
+                    .expect("expect argument ends without '\\'")
+                    .1
+                    .trim();
+
+                line.push_str(l);
+
+                // We need the last line number of argument
+                pos += 1;
+            }
+
+            debug!(self.logger, "argument_after_cleanup: {}", line);
+
+            // Start reading argument
+            // Init data
+            self.data = TestData::default();
+
+            // Save `line` information for error/debug message usage
+            self.data.pos = format!("{} : L{}", self.source_name.as_path().display(), pos + 1);
+
+            let (cmd, cmd_args) = parse_line(line.as_str(), &self.logger)?;
+
+            if cmd.is_empty() {
+                bail!("cmd must not be empty");
+            }
+
+            debug!(self.logger, "cmd: {}, cmd_args: {:?}", cmd, cmd_args,);
+
+            self.data.cmd = cmd;
+            self.data.cmd_args = cmd_args;
+
+            loop {
+                let line = self.scanner.next();
+                if line.is_none() {
+                    bail!("testdata is not complete, we expected '----' after command");
+                }
+
+                let line = line.unwrap().1;
+
+                if line == "----" {
+                    self.read_expected();
+                    return Ok(true);
+                } else if !line.is_empty() {
+                    bail!("expected '----', found '{}'", line);
+                }
+            }
+        }
+    }
+
+    // check whether there is two separator or one
+    fn read_expected(&mut self) {
+        // after reading first separator
+        // we expected two input
+        // (1) second separator
+        // (2) non empty output
+
+        // We need to add an end line for each line in order to distinguish the exact output,
+        // otherwise we may consider an unexpected output combination to be correct
+
+        if let Some((_, line)) = self.scanner.next() {
+            if line == "----" {
+                loop {
+                    let mut line = self.scanner.next().unwrap().1.trim().to_string();
+                    if line == "----" {
+                        let mut line2 = self.scanner.next().unwrap().1.trim().to_string();
+                        if line2 == "----" {
+                            let line3 = self.scanner.next().unwrap().1.trim().to_string();
+                            assert!(
+                                line3.is_empty(),
+                                "we expected an empty line after second separator, found '{}'",
+                                line3
+                            );
+                            break;
+                        }
+                        if !line2.is_empty() {
```

Comment:
Why check if empty?

Reply:
That is to skip blank line, only add "\n" to others.
Currently, I skip the blank line to expected value.
e.g.
```

a=1
```
`expected string` = "a=1\n"

If we don't skip blank line, 
we will expect to get `expected string`  = "\na=1\n".

Reply:
What's the golang behavior?

Reply:
I think they do the latter, I will follow it.